lms/djangoapps/course_api/views.py

```diff
@@ -201,9 +210,24 @@ def get_queryset(self):
         if not form.is_valid():
             raise ValidationError(form.errors)
 
-        return list_courses(
+        db_courses = list_courses(
             self.request,
             form.cleaned_data['username'],
             org=form.cleaned_data['org'],
             filter_=form.cleaned_data['filter_'],
         )
+
+        if not settings.FEATURES['ENABLE_COURSEWARE_SEARCH'] or not form.cleaned_data['search_term']:
+            return db_courses
+
+        search_courses = search.api.course_discovery_search(
+            form.cleaned_data['search_term'],
+            size=self.results_size_infinity,
+        )
+
+        search_courses_ids = {course['data']['id']: True for course in search_courses['results']}
```

Comment:
This seems straightforward, but I am still trying to understand the need to pull in both the DB data and the search results and compare the two.  Do the search results only contain the course ids that match the search?

Reply:
@staubina Good question! The need to pull from both DBs comes from the fact that all the information we need is actually stored in the CourseOverviews (and Mongo):

 - **Does course support mobile**
 - **Access controls**, whether the user has access to a course
 - Dates
 - Images
 - Actual titles
 - Etc.

But of course, we can do proper full-text search via MySQL (and Mongo), so the full text search is being done via ElasticSearch (using `edx-search`).

If `edx-search` had all the information we need mainly: access controls and whether the course is on mobile, then we can rely exclusively on ElasticSearch.

The only viable alternative scenario is the following:

 - Pull the courses from ElasticSearch that matches the search
 - For each course do the following:
   - Make sure it supports mobile (based on queryparam)
   - Make sure the user has access to it
   - Keep doing so, until we grab enough courses for that page
 - If `page=2` is requested, then we'd have to do the above for both pages

The above solution is both more complex to understand and maintain, and _might_ only provide noticeable performance on the first pages. Subsequent pages could have a worse performance than the implemented solution in this PR (#16757).

What do you think?

Reply:
I think this is acceptable, I hope that we can move to using Course Discovery in the not to distant future. @OmarIthawi 

I have approval from the Product team to continue with this.  I am going to finish my review in the next few days. 

One request can you ask the mobile engineers working on the Native implementation to consider making it as simple as possible to add in a new search location like Course discovery in the future?  

I look forward to those PRs as well.

Reply:
Thanks @staubina!

> One request can you ask the mobile engineers working on the Native implementation to consider making it as simple as possible to add in a new search location like Course discovery in the future?

@jaimeohm please checkout :point_up_2:

Reply:
@staubina Hello! I'm working on the native implementation. Could you clarify what you mean by this? I'm adding the URL for search as a configuration flag. Does this offer the flexibility you're hoping for? 

Thanks for the heads up, @OmarIthawi.

Reply:
I think that the configurable URL should work.  I may be overthinking it at this point.  

If there is a PR in the future we can discuss it there. The configurable URL should work to start and we will add any data interpretation layers we need for course-discovery if and when that comes up.

Reply:
@jaimeohm ^